### ISSUE TITLE
CUDAOutOfMemoryError for duplex with 3080ti (12Gb) 

### ISSUE THREAD

**Issue by olawa** [nanoporetech/dorado#57]
Hi,

duplex fails with OOM as shown below, is it possible to reduce memory footprint like with chunks_per_runner in guppy?

```
terminate called after throwing an instance of 'c10::CUDAOutOfMemoryError'
  what():  CUDA out of memory. Tried to allocate 1.95 GiB (GPU 0; 11.77 GiB total capacity; 4.51 GiB already allocated; 1.47 GiB free; 8.38 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
Exception raised from malloc at ../c10/cuda/CUDACachingAllocator.cpp:578 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x3e (0x7f12ef63720e in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libc10.so)
frame #1: <unknown function> + 0x1667f (0x7f12ef6a167f in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libc10_cuda.so)
frame #2: <unknown function> + 0x46528 (0x7f12ef6d1528 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libc10_cuda.so)
frame #3: <unknown function> + 0x46752 (0x7f12ef6d1752 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libc10_cuda.so)
frame #4: at::detail::empty_generic(c10::ArrayRef<long>, c10::Allocator*, c10::DispatchKeySet, c10::ScalarType, c10::optional<c10::MemoryFormat>) + 0x7bf (0x7f12f0ee46af in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cpu.so)
frame #5: at::detail::empty_cuda(c10::ArrayRef<long>, c10::ScalarType, c10::optional<c10::Device>, c10::optional<c10::MemoryFormat>) + 0x115 (0x7f13098c3ca5 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cuda_cpp.so)
frame #6: at::detail::empty_cuda(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) + 0x31 (0x7f13098c3f01 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cuda_cpp.so)
frame #7: at::detail::empty_cuda(c10::ArrayRef<long>, c10::TensorOptions const&) + 0x10f (0x7f13098c406f in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cuda_cpp.so)
frame #8: <unknown function> + 0x2d720db (0x7f12c22340db in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cuda_cu.so)
frame #9: <unknown function> + 0x2e289c6 (0x7f12c22ea9c6 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cuda_cu.so)
frame #10: at::meta::structured_mm::meta(at::Tensor const&, at::Tensor const&) + 0x179 (0x7f12f11f6019 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cpu.so)
frame #11: <unknown function> + 0x2d7ad2a (0x7f12c223cd2a in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cuda_cu.so)
frame #12: <unknown function> + 0x2d7adc0 (0x7f12c223cdc0 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cuda_cu.so)
frame #13: at::_ops::mm::call(at::Tensor const&, at::Tensor const&) + 0xc4 (0x7f12f1af5084 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cpu.so)
frame #14: <unknown function> + 0x1a7dc05 (0x7f12f11f8c05 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cpu.so)
frame #15: at::native::matmul(at::Tensor const&, at::Tensor const&) + 0x40 (0x7f12f11f90e0 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cpu.so)
frame #16: <unknown function> + 0x273c1b0 (0x7f12f1eb71b0 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cpu.so)
frame #17: at::_ops::matmul::call(at::Tensor const&, at::Tensor const&) + 0xc4 (0x7f12f1bc4a84 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cpu.so)
frame #18: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x4fb255]
frame #19: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x5205f1]
frame #20: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x520f1e]
frame #21: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x58932c]
frame #22: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x588233]
frame #23: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x5872e1]
frame #24: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x583b26]
frame #25: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x571bae]
frame #26: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x56bdd3]
frame #27: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x56359c]
frame #28: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x589d4e]
frame #29: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x5886b7]
frame #30: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x5879d9]
frame #31: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x585fba]
frame #32: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x5719ea]
frame #33: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x65a50f]
frame #34: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x659dc9]
frame #35: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x65ce78]
frame #36: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x65cdba]
frame #37: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x65cd29]
frame #38: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x65ccb6]
frame #39: /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/dorado() [0x65cc5a]
frame #40: <unknown function> + 0x145a0 (0x7f13187b55a0 in /home/ola/data/dorado-0.1.0+4b0e9a6-Linux/bin/../lib/libtorch_cuda.so)
frame #41: <unknown function> + 0x8609 (0x7f12bed34609 in /lib/x86_64-linux-gnu/libpthread.so.0)
frame #42: clone + 0x43 (0x7f12be901133 in /lib/x86_64-linux-gnu/libc.so.6)

Aborted (core dumped)
```

**Comment by gwl2** [comment]
I have the same/similar issue  on my 3080 12GB. Simplex basecalling in SUP mode with the 4.1.0 model works as expected. 
However, dorado wouldn't start duplex basecalling at all due allocation of memory issue.
When I stopped the guppy server (running on the GPU) from MinKNOW, it started but crashed after 200 reads. Afterwards it wouldn't start anymore, always ended up with an memory allocation error. See the logs below.
Of note, I have an AMD CPU, I think it was discussed in another discussion here, that this might be related to the type of CPU (same card on an intel CPU was not giving an memory allocation issue).

**That's the output of the trial that stopped after 200 reads (see below for the other):**
./dorado duplex dna_r10.4.1_e8.2_260bps_sup@v4.1.0 /data/RV_exploration/KP02/pod5_bc06/ --pairs /data/RV_exploration/KP02/pod5_bc06/pairs_from_sam/pair_ids_filtered.txt --emit-fastq > /data/RV_exploration/KP02/pod5_bc06/duplex_orig.fastq
[2023-03-29 15:27:39.128] [info] > Loading pairs file
[2023-03-29 15:27:39.129] [info] > Pairs file loaded
[2023-03-29 15:27:41.572] [info] > Starting Stereo Duplex pipeline
> Reads processed: 200terminate called after throwing an instance of 'c10::OutOfMemoryError'
  what():  CUDA out of memory. Tried to allocate 1.95 GiB (GPU 0; 11.74 GiB total capacity; 4.51 GiB already allocated; 1.43 GiB free; 8.38 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
Exception raised from malloc at ../c10/cuda/CUDACachingAllocator.cpp:681 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7fcaa4e98457 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libc10.so)
frame #1: <unknown function> + 0x4442d (0x7fcaa4f3642d in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libc10_cuda.so)
frame #2: <unknown function> + 0x45158 (0x7fcaa4f37158 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libc10_cuda.so)
frame #3: <unknown function> + 0x453c2 (0x7fcaa4f373c2 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libc10_cuda.so)
frame #4: at::detail::empty_generic(c10::ArrayRef<long>, c10::Allocator*, c10::DispatchKeySet, c10::ScalarType, c10::optional<c10::MemoryFormat>) + 0xabf (0x7fcaa692627f in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #5: at::detail::empty_cuda(c10::ArrayRef<long>, c10::ScalarType, c10::optional<c10::Device>, c10::optional<c10::MemoryFormat>) + 0x111 (0x7fcac0453c51 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cpp.so)
frame #6: at::detail::empty_cuda(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) + 0x31 (0x7fcac0453f21 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cpp.so)
frame #7: at::detail::empty_cuda(c10::ArrayRef<long>, c10::TensorOptions const&) + 0x10f (0x7fcac045408f in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cpp.so)
frame #8: <unknown function> + 0x2d1622b (0x7fca7299c22b in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cu.so)
frame #9: <unknown function> + 0x2dd5b06 (0x7fca72a5bb06 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cu.so)
frame #10: at::meta::structured_mm::meta(at::Tensor const&, at::Tensor const&) + 0x124 (0x7fcaa6d8f024 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #11: <unknown function> + 0x2d1ebf5 (0x7fca729a4bf5 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cu.so)
frame #12: <unknown function> + 0x2d1ec80 (0x7fca729a4c80 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cu.so)
frame #13: at::_ops::mm::call(at::Tensor const&, at::Tensor const&) + 0x161 (0x7fcaa78500f1 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #14: <unknown function> + 0x1dabff5 (0x7fcaa6d9dff5 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #15: at::native::matmul(at::Tensor const&, at::Tensor const&) + 0x49 (0x7fcaa6d9e499 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #16: <unknown function> + 0x2d6ea40 (0x7fcaa7d60a40 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #17: at::_ops::matmul::call(at::Tensor const&, at::Tensor const&) + 0xc3 (0x7fcaa794c283 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #18: ./dorado() [0x4d1597]
frame #19: ./dorado() [0x51d583]
frame #20: ./dorado() [0x51d853]
frame #21: ./dorado() [0x4f0573]
frame #22: ./dorado() [0x4f0680]
frame #23: ./dorado() [0x51cd50]
frame #24: ./dorado() [0x4f041d]
frame #25: ./dorado() [0x554fe0]
frame #26: <unknown function> + 0x145a0 (0x7fcacf9145a0 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda.so)
frame #27: <unknown function> + 0x8609 (0x7fca6f045609 in /lib/x86_64-linux-gnu/libpthread.so.0)
frame #28: clone + 0x43 (0x7fca6ec12133 in /lib/x86_64-linux-gnu/libc.so.6)

Aborted (core dumped)


**Nothing changed just reexecuted the command:**
/dorado duplex dna_r10.4.1_e8.2_260bps_sup@v4.1.0 /data/RV_exploration/KP02/pod5_bc06/ --pairs /data/RV_exploration/KP02/pod5_bc06/pairs_from_sam/pair_ids_filtered.txt --emit-fastq -b 192 > /data/RV_exploration/KP02/pod5_bc06/duplex_orig.fastq
[2023-03-29 15:32:15.044] [info] > Loading pairs file
[2023-03-29 15:32:15.044] [info] > Pairs file loaded
[2023-03-29 15:32:17.245] [info] > Starting Stereo Duplex pipeline
terminate called after throwing an instance of 'c10::OutOfMemoryError'
  what():  CUDA out of memory. Tried to allocate 1.95 GiB (GPU 0; 11.74 GiB total capacity; 5.01 GiB already allocated; 1.42 GiB free; 8.46 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
Exception raised from malloc at ../c10/cuda/CUDACachingAllocator.cpp:681 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7fc04f5c8457 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libc10.so)
frame #1: <unknown function> + 0x4442d (0x7fc04f66642d in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libc10_cuda.so)
frame #2: <unknown function> + 0x45158 (0x7fc04f667158 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libc10_cuda.so)
frame #3: <unknown function> + 0x453c2 (0x7fc04f6673c2 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libc10_cuda.so)
frame #4: at::detail::empty_generic(c10::ArrayRef<long>, c10::Allocator*, c10::DispatchKeySet, c10::ScalarType, c10::optional<c10::MemoryFormat>) + 0xabf (0x7fc05105627f in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #5: at::detail::empty_cuda(c10::ArrayRef<long>, c10::ScalarType, c10::optional<c10::Device>, c10::optional<c10::MemoryFormat>) + 0x111 (0x7fc06ab83c51 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cpp.so)
frame #6: at::detail::empty_cuda(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) + 0x31 (0x7fc06ab83f21 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cpp.so)
frame #7: at::detail::empty_cuda(c10::ArrayRef<long>, c10::TensorOptions const&) + 0x10f (0x7fc06ab8408f in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cpp.so)
frame #8: <unknown function> + 0x2d1622b (0x7fc01d0cc22b in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cu.so)
frame #9: <unknown function> + 0x2dd5b06 (0x7fc01d18bb06 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cu.so)
frame #10: at::meta::structured_mm::meta(at::Tensor const&, at::Tensor const&) + 0x124 (0x7fc0514bf024 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #11: <unknown function> + 0x2d1ebf5 (0x7fc01d0d4bf5 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cu.so)
frame #12: <unknown function> + 0x2d1ec80 (0x7fc01d0d4c80 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda_cu.so)
frame #13: at::_ops::mm::call(at::Tensor const&, at::Tensor const&) + 0x161 (0x7fc051f800f1 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #14: <unknown function> + 0x1dabff5 (0x7fc0514cdff5 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #15: at::native::matmul(at::Tensor const&, at::Tensor const&) + 0x49 (0x7fc0514ce499 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #16: <unknown function> + 0x2d6ea40 (0x7fc052490a40 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #17: at::_ops::matmul::call(at::Tensor const&, at::Tensor const&) + 0xc3 (0x7fc05207c283 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cpu.so)
frame #18: ./dorado() [0x4d1597]
frame #19: ./dorado() [0x51d583]
frame #20: ./dorado() [0x51d853]
frame #21: ./dorado() [0x4f0573]
frame #22: ./dorado() [0x4f0680]
frame #23: ./dorado() [0x51cd50]
frame #24: ./dorado() [0x4f041d]
frame #25: ./dorado() [0x554fe0]
frame #26: <unknown function> + 0x145a0 (0x7fc07a0445a0 in /home/nanopore/software/dorado-0.2.1-linux-x64/bin/../lib/libtorch_cuda.so)
frame #27: <unknown function> + 0x8609 (0x7fc019775609 in /lib/x86_64-linux-gnu/libpthread.so.0)
frame #28: clone + 0x43 (0x7fc019342133 in /lib/x86_64-linux-gnu/libc.so.6)

Aborted (core dumped)

**Comment by caonetto** [comment]
Same issue here, cant finish duplex basecalling on a Geforce RTX 3080 (10gb). Crashes after a certain number of reds throwing the CUDA out of memory error. When its running, nvidia-smi displays that 9838MiB / 10240MiB are being used, and I am using a -b of 64. Should I go even lower?

**Comment by opplatek** [comment]
Hi everyone. Same issue here with NVIDIA GeForce RTX 2080 Ti  (12 GB), `nvidia-driver-530` driver, `dorado-0.2.4-linux-x64` on Ubuntu 18.04. Simplex basecalling with `dna_r10.4.1_e8.2_400bps_sup@v4.1.0` model worked fine, duplex crashed almost immediately after `[info] > Starting Stereo Duplex pipeline` due to the same GPU memory error as mentioned above.
I started with `-b 160` and ended with `-b 80`. All the other multiples of 16 in between crashed at different stages of basecalling. Basecalling finished without crashing at `-b 80`. I tried to launch it several times after the successful run, and it seems it crashed with a different `-b xx` every time.
